examples/vpc-cni-custom-networking/main.tf

```diff
@@ -0,0 +1,263 @@
+    # Version 1.9.0 or later (for version 1.20 or earlier clusters or 1.21 or later clusters configured for IPv4)
+    # or 1.10.1 or later (for version 1.21 or later clusters configured for IPv6) of the Amazon VPC CNI for prefix delegation
+    addon_version     = data.aws_eks_addon_version.latest["vpc-cni"].version
+    resolve_conflicts = "OVERWRITE"
+  }
+
+  tags = local.tags
+
+  depends_on = [
+    # Modify VPC CNI ahead of addons
+    null_resource.kubectl_set_env
+  ]
+}
+
+data "aws_eks_addon_version" "latest" {
+  for_each = toset(["vpc-cni"])
+
+  addon_name         = each.value
+  kubernetes_version = module.eks_blueprints.eks_cluster_version
+  most_recent        = true
+}
+
+#---------------------------------------------------------------
+# Modify VPC CNI deployment
+#---------------------------------------------------------------
+
+locals {
+  kubeconfig = yamlencode({
+    apiVersion      = "v1"
+    kind            = "Config"
+    current-context = "terraform"
+    clusters = [{
+      name = module.eks_blueprints.eks_cluster_id
+      cluster = {
+        certificate-authority-data = module.eks_blueprints.eks_cluster_certificate_authority_data
+        server                     = module.eks_blueprints.eks_cluster_endpoint
+      }
+    }]
+    contexts = [{
+      name = "terraform"
+      context = {
+        cluster = module.eks_blueprints.eks_cluster_id
+        user    = "terraform"
+      }
+    }]
+    users = [{
+      name = "terraform"
+      user = {
+        token = data.aws_eks_cluster_auth.this.token
+      }
+    }]
+  })
+}
+
+resource "null_resource" "kubectl_set_env" {
+  triggers = {}
+
+  provisioner "local-exec" {
+    interpreter = ["/bin/bash", "-c"]
+    environment = {
+      KUBECONFIG = base64encode(local.kubeconfig)
+    }
+
+    # Reference https://aws.github.io/aws-eks-best-practices/reliability/docs/networkmanagement/#cni-custom-networking
+    # Reference https://docs.aws.amazon.com/eks/latest/userguide/cni-increase-ip-addresses.html
+    command = <<-EOT
+      # Custom networking
+      kubectl set env daemonset aws-node -n kube-system AWS_VPC_K8S_CNI_CUSTOM_NETWORK_CFG=true --kubeconfig <(echo $KUBECONFIG | base64 --decode)
+      kubectl set env daemonset aws-node -n kube-system ENI_CONFIG_LABEL_DEF=failure-domain.beta.kubernetes.io/zone --kubeconfig <(echo $KUBECONFIG | base64 --decode)
+
+
+      # Prefix delegation
+      kubectl set env daemonset aws-node -n kube-system ENABLE_PREFIX_DELEGATION=true --kubeconfig <(echo $KUBECONFIG | base64 --decode)
+      kubectl set env daemonset aws-node -n kube-system WARM_PREFIX_TARGET=1 --kubeconfig <(echo $KUBECONFIG | base64 --decode)
+    EOT
+  }
+}
+
+#---------------------------------------------------------------
+# VPC-CNI Custom Networking ENIConfig
+#---------------------------------------------------------------
+
+resource "kubectl_manifest" "eni_config" {
```

Comment:
does it matter the order of this resource? should the manifest come first right after the cluster is created AND before the addons?

Reply:
yes order matters - the VPC CNI needs to be updated and the ENIConfig present before nodes are ready. However, since custom networking should be set at cluster creation, the order of operations work out as is naturally